blog/content/post/2024-11-26-measured-boot-improvements.md

```diff
@@ -0,0 +1,222 @@
+are already upstream). Besides the TPM ACPI table, coreboot creates an event
+log using only one hashing algorithm:
+
+- SHA1 - default for TPM1.2
+- SHA256 - default for TPM2.0
+
+TPM2.0 log format supports agile measurements, i.e. there may be multiple
+measurements of the same component using multiple hashing algorithms. For
+simplicity, coreboot uses only one currently. The choice of the hashing
+algorithm must also be passed to the UEFI Payload. It is done by [specifying a
+mask of supported hashing algorithms at UEFI Payload build
+time](https://github.com/Dasharo/coreboot/pull/517/files#diff-58bb01776a00af5a5d61de00d6ea3008bcd67177fb7991fb41e8976c6b4d046a).
+
+Now that all the pieces are in place, we can enjoy having a consistent and
+complete TPM event log in the Dasharo, effectively solving the [issue of not
+reproducible PCR2
+value](https://github.com/Dasharo/dasharo-issues/issues/455).
+
+## Summary
+
+TPM event log integrity is an excellent achievement for our Dasharo team. While we
+did an enormous work to improve it, there are still some aspects not covered
+by current implementation:
+
+1. What if the PCR banks are changed?
+
+   The EDKII TPM2 setup menu allows changing PCR banks, effectively
+   influencing what measurements (precisely hash algorithms used to measure
+   firmware components) may be recorded in the TPM. coreboot is hardcoded to
+   use only one hashing algorithm (fixed at build). It may result in PCR banks
+   not being properly extended or the PCR values being entirely out of sync
+   with the TPM event log. coreboot should query the TPM for currently enabled
+   PCR banks and use all algorithms enabled at the time. That, of course, may
+   have various implications:
+
+    - Pre-RAM measurements are saved in Cache-as-RAM (CAR) memory, which is
+      limited. Calculating more hashes will cause the CAR memory to shrink,
+      which effectively limits the amount of cache for silicon initialization
+      code. It may not be a big deal for modern platforms with big caches, but
+      rather for the old ones.
+    - coreboot supports only one algorithm in a single event log entry.
+      Support for an agile format would have to be implemented.
+
+2. What about pre-reset vector measurements like Boot Guard measured boot?
+
+   Intel Boot Guard is a hardware Root of Trust feature that allows the
+   firmware to be verified (and optionally measured) before the CPU is
+   released from reset. It guarantees the integrity and reliable measurement
+   of the critical early boot code. Currently, coreboot ignores the fact that
+   something may have extended TPM PCRs with measurements before coreboot
+   initialized the TPM or started logging measurements. Support for Boot
+   Guard-created measurements must also be added to keep the PCR values
+   consistent with the TPM event log.
+
+Improving the security of the firmware is a process. What is important is to
```

Comment:
This is summary IMHO

Reply:
cff1006b43a13b63a542f10eed3a227911bce453